app/lib/user_annotation_service.rb

```diff
@@ -0,0 +1,102 @@
+        # models/user_annotation.rb to remove `subsample_annotation`.  Jon
+        # and Eric investigated on 2021-01-22 and determined that there are no
+        # known use cases where `subsample_annotation` != `loaded_annotation`.
+        subsample_annotation = loaded_annotation
+
+        # Method call to create the user data arrays for this annotation
+        user_annotation.initialize_user_data_arrays(user_data_arrays_attributes, subsample_annotation, subsample_threshold, loaded_annotation)
+
+        # Reset the annotations in the dropdowns to include this new annotation
+        cluster_annotations = ClusterVizService.load_cluster_group_annotations(study, cluster, current_user)
+
+        message = "User Annotation: '#{user_annotation.name}' successfully saved. You may now view this annotation via the annotations dropdown."
+        status = 200 # Success
+      else
+        # If there was an error saving, alert the user something broke
+        cluster_annotations = ClusterVizService.load_cluster_group_annotations(study, cluster, current_user)
+        message = 'The following errors prevented the annotation from being saved: ' + user_annotation.errors.full_messages.join(',')
+        Rails.logger.error "Creating user annotation of params: #{user_annotation_params}, unable to save user annotation with errors #{user_annotation.errors.full_messages.join(', ')}"
+        status = 400  # Bad request
+      end
+      [message, cluster_annotations, status]
+
+    # Handle other errors in saving user annotation
+    rescue Mongoid::Errors::InvalidValue => e
+      error_context = ErrorTracker.format_extra_context(study, {params: log_params})
+      ErrorTracker.report_exception(e, current_user, error_context)
+      # If an invalid value was somehow passed through the form, and couldn't save the annotation
+      cluster_annotations = ClusterVizService.load_cluster_group_annotations(study, cluster, current_user)
+      message = 'The following errors prevented the annotation from being saved: ' + 'Invalid data type submitted. (' + e.problem + '. ' + e.resolution + ')'
+      Rails.logger.error "Creating user annotation of params: #{user_annotation_params}, invalid value of #{e.message}"
+      [message, cluster_annotations, 400] # Bad request
+
+    rescue NoMethodError => e
+      error_context = ErrorTracker.format_extra_context(study, {params: log_params})
+      ErrorTracker.report_exception(e, current_user, error_context)
+      # If something is nil and can't have a method called on it, respond with an alert
+      cluster_annotations = ClusterVizService.load_cluster_group_annotations(study, cluster, current_user)
+      message = 'The following errors prevented the annotation from being saved: ' + e.message
+      Rails.logger.error "Creating user annotation of params: #{user_annotation_params}, no method error #{e.message}"
+      [message, cluster_annotations, 500] # Internal server error
+
+    rescue => e
```

Comment:
what does this add beyond our current default API controller exception handling?  I think the error message itself belongs in the controller or the UI, not here.  I think this block should be removed and, to the extent it isn't duplicative of our existing error-handling code, reallocate it to the controller

Reply:
Done -- latest code refines this aspect of the old `site_controller.rb` handling.